toolchain/semantics/semantics_ir_formatter.cpp

```diff
@@ -0,0 +1,584 @@
+                                   ? semantics_ir.GetString(fn.name_id).str()
+                                   : "");
+      CollectNamesInBlock(fn_scope, fn.param_refs_id);
+      for (auto block_id : fn.body_block_ids) {
+        AddBlockLabel(fn_scope, block_id,
+                      block_id == fn.body_block_ids.front() ? "entry" : "");
+        CollectNamesInBlock(fn_scope, block_id);
+      }
+    }
+  }
+
+  // Returns the scope index corresponding to a function.
+  auto GetScopeFor(SemanticsFunctionId fn_id) -> ScopeIndex {
+    return ScopeIndex(fn_id.index + 1);
+  }
+
+  // Returns the IR name to use for a function.
+  auto GetNameFor(SemanticsFunctionId fn_id) -> std::string {
+    if (!fn_id.is_valid()) {
+      return "invalid";
+    }
+    return GetScopeInfo(GetScopeFor(fn_id)).name;
+  }
+
+  // Returns the IR name to use for a node, when referenced from a given scope.
+  auto GetNameFor(ScopeIndex scope_idx, SemanticsNodeId node_id)
+      -> std::string {
+    if (!node_id.is_valid()) {
+      return "invalid";
+    }
+
+    // Check for a builtin.
+    if (node_id.index < SemanticsBuiltinKind::ValidCount) {
+      return SemanticsBuiltinKind::FromInt(node_id.index).label().str();
+    }
+
+    auto& [node_scope, node_name] = nodes[node_id.index];
+    if (node_name.empty()) {
+      // This should not happen in valid IR.
+      return "<noderef " + llvm::itostr(node_id.index) + ">";
+    }
+    if (node_scope == scope_idx) {
+      return node_name;
+    }
+    return GetScopeInfo(node_scope).name + "." + node_name;
+  }
+
+  // Returns the IR name to use for a label, when referenced from a given scope.
+  auto GetLabelFor(ScopeIndex scope_idx, SemanticsNodeBlockId block_id)
+      -> std::string {
+    if (!block_id.is_valid()) {
+      return "!invalid";
+    }
+
+    auto& [label_scope, label_name] = labels[block_id.index];
+    if (label_name.empty()) {
+      // This should not happen in valid IR.
+      return "<nodeblockref " + llvm::itostr(block_id.index) + ">";
+    }
+    if (label_scope == scope_idx) {
+      return label_name;
+    }
+    return GetScopeInfo(label_scope).name + "." + label_name;
+  }
+
+ private:
+  // A space in which unique names can be allocated.
+  struct Namespace {
+    llvm::StringSet<> allocated;
+    int unnamed_count = 0;
+
+    auto AllocateName(const NodeNamer& namer, ParseTree::Node node,
+                      std::string name = "") -> std::string {
+      // Use the given name if it's available.
+      if (!name.empty() && allocated.insert(name).second) {
+        return name;
+      }
+
+      // Append location information to try to disambiguate.
+      if (node.is_valid()) {
+        auto token = namer.parse_tree_.node_token(node);
+        name += ".L";
+        name += llvm::itostr(namer.tokenized_buffer_.GetLineNumber(token));
+        if (allocated.insert(name).second) {
+          return name;
+        }
+        name += ".";
```

Comment:
I think it's useful to distinguish the column number from a non-location disambiguator.

I know it makes the name longer, but maybe:
```suggestion
        name += ".c";
```

I also find `c` a bit easier to separate from the number than `C`, and I'd probably use `l` for consistency, although there both `l` and `L` are hard to visually separate. Somewhat minor though.

Reply:
Format updated per in-person discussion.